### ISSUE TITLE
False negatives are very high - yolov3

### ISSUE THREAD

**Issue by kmsravindra** [AlexeyAB/darknet#1739]
@AlexeyAB , I have 1920x1080 images that I am using for training with Yolo v3 model. I have kept my training network width at 608x608 and trained for around 7000 iterations. The object width and height typically vary between 40wx40h to 300wx300h ( these are approx to just give you a feel of the object sizes). This is a single class detection with ~6000 positive images(with object present) and ~6000 background images(with no object). So there are total of 12000+ images that I am training upon. All of them are 1920x1080 images.

Wen I test the model on the hold out test set ( which is also at 1920x1080 size), I am seeing that the number of false negatives to be very high (almost 90%) using a 608x608 detection network size.
So, when I increase the detection network size to 1280x1280, the false negatives reduced but they are still around 80% of the total true positives. The number of false positives and true positives are substantially very very low.
 
And this false negative numbers are always high however many ways I trained my model. 

Could you please suggest - 
1. Are the training network sizes and detection network sizes good for the input image size that I have?
2. Should I do any config changes ( I am already using random=1 and the typical learning rate params)?
3. Do you think I need to increase my positive images data to include more variety of objects?
4. Any other changes / suggestions would be very valuable as I have spent lot of time varying several parameters but is of not much use.

Please suggest.

thanks!

**Comment by AlexeyAB** [comment]
@kmsravindra Hi,

1. How did you get your Test-dataset? Did you just randomly separate initial dataset to the 2 datasets: Training (80%) and Test (20%)?

2. Can you show screenshots of windows that will be generated for Training (`train=train.txt` in obj.data) and Test datasets (`train=test.txt` in obj.data)?
`./darknet detector calc_anchors data/obj.data -num_of_clusters 9 -width 608 -height 608 -show`

**Comment by kmsravindra** [comment]
@AlexeyAB , Hi, Thanks

1. The test data set is completely a new video image sequence whose frames were never part of the training model. This is how the production scenario in our case would be. Training images were taken from video1 to video50 let's say then the testing images are from 51 to 60.

2. You are correct that my anchors are not consistent across train data and test data - But I do not know how to circumvent such issue give my above case that the test data can be completely new...Please suggest on what can be done in such cases to take care of this inconsistency issues from video to video?
Below is the information you have asked for - 

Train data:
=======
paperspace@psebyt5vu:~/Documents/darknet-master/build/darknet/x64$ ./darknet detector calc_anchors data/obj.data -num_of_clusters 9 -width 608 -height 608 --
show  num_of_clusters = 9, width = 608, height = 608 
 read labels from 13175 images 
 loaded 	 image: 13170 	 box: 5542 all loaded. 

calculating k-means++ ...
 
avg IoU = 74.80 %

Saving anchors to the file: anchors.txt 

anchors = 40.8708,60.2445, 55.2498,98.0356, 65.5211,137.6225, 109.8926,120.2942, 93.6058,183.5968, 107.6155,248.0035, 151.3014,191.1671, 170.5979,309.8748, 162.9289,428.1051

Test data:
=======

paperspace@psebyt5vu:~/Documents/darknet-master/build/darknet/x64$ ./darknet detector calc_anchors data/obj.data -num_of_clusters 9 -width 608 -height 608 --
show num_of_clusters = 9, width = 608, height = 608 
 
read labels from 460 images  loaded 	 image: 460 	 box: 362 all loaded. 

calculating k-means++ ...
 
avg IoU = 80.82 % 

Saving anchors to the file: anchors.txt 
anchors = 58.8718,46.7510, 41.6118,70.1583, 76.6400,66.9456, 60.3822,89.6283, 75.3112,117.6099, 64.2508,151.3216, 112.1679,162.3344, 104.7607,217.1381, 146.9335,309.0665

**Comment by AlexeyAB** [comment]
@kmsravindra Hi,

1. Can you show screenshot of cloud of points by using `calc_anchors` with the flag `-show`?

2. What mAP can you get on the Training dataset? `valid=train.txt` in the obj.data file.

3. Most likely the images of objects in the Training-set are completely different from the images in the Test-set. As if you were training for the 1 class “Transport”, but there were only “Cars” in the Training-set, and you would try to detect “Airs” in the Test-set.

4. Also try to find several false-negative examples from Test-set, then try to find the most similary objects in the Training-set, and try to describe the main difference between these objects: color, size, aspect-ratio, shape, details, different size, ...

5. I recommend you to increase your Training dataset. And follow this rule: 

> desirable that your training dataset include images with objects at diffrent: scales, rotations, lightings, from different sides, on different backgrounds

6. I recommend you to train at least 12 000 - 15 000 iterations, if you have 12 000 images.

6. Also you can try to set in your cfg-file: `saturation=2.5 exposure=2.5 hue=0.3` in the [net]-section and `jitter=0.4 random=1` in the each of 3 [yolo]-layers, and use. Then train about **120 000 iterations.** It will use more agressive data augmentation, if you can't get more different Training images.

**Comment by kmsravindra** [comment]
@AlexeyAB , Thanks for your response. Here is the information:
1. 
train clusters:
![trainclusters](https://user-images.githubusercontent.com/3147499/46781420-8e36b480-ccef-11e8-98c8-1deea149e6f4.png)

test clusters:
![test-clusters](https://user-images.githubusercontent.com/3147499/46781434-94c52c00-ccef-11e8-8201-7de43a3980c5.png)

2. mAP on training dataset is quite good! at just 2700 iterations as below :

detections_count = 16372, unique_truth_count = 5542  
class_id = 0, name = polyp, 72   ap = 90.23 % 
 for thresh = 0.25, precision = 0.87, recall = 0.94, F1-score = 0.91 
 for thresh = 0.25, TP = 5221, FP = 756, FN = 321, average IoU = 67.74 % 

 mean average precision (mAP) = 0.902264, or 90.23 % 

3 and 4. The images and objects that we are dealing with are inside the human intestine (colonoscopy videos) and objects are early cancerous growths, small lumps etc., Some of them are mostly blended with the background and some of them form odd lumps...Our plan is to detect tough ones that the physician finds blended into the skin...Given this nature of dataset, what do you recommend? Assume we can collect several colonoscopy HD videos without limit...so data is not a problem. We are extracting frames from each video that constitute our training dataset. Testing / inference of the model will have to be done on a new video. Also just as a note, since there are several background frames from each video, I am doing a random sampling of negative images from hundreds of videos to gather equal number of images as positive images. So the dataset of 12000+ images includes 6000+ positive images and 6000+ negative randomly sampled images.

5. Some of our data comes inherently augmented as deal with video sequence of frames with camera probing the lump from several angles...Nevertheless, we are using all the augmentations of yolov3 defaults.

6. Sure, I could do that...but I already saw overfitting happening around 4000 to 5000 iterations. For example the mAP that I have above is at just 2700 iterations and loss was showing up as 0.13+ which I believe is quite insignificant

7. As I mentioned some of the data aug happens implicitly but will try this aggressive option as well...Training for 120000 iterations might take a week / 10 days long with the machine that we have currently. If you think this is really going to impact then I can give it a try.


Given this context, do you have any suggestions as to how we should prepare the data otherwise OR is YOLOv3 good enough when there are blended objects into the background? OR should we do something else? Your suggestions will be highly helpful!

thanks a lot!

**Comment by AlexeyAB** [comment]
@kmsravindra 

> mAP on training dataset is quite good! at just 2700 iterations as below :
> detections_count = 16372, unique_truth_count = 5542
> class_id = 0, name = polyp, 72 ap = 90.23 %
> for thresh = 0.25, precision = 0.87, recall = 0.94, F1-score = 0.91
> for thresh = 0.25, TP = 5221, FP = 756, FN = 321, average IoU = 67.74 %
> 
> mean average precision (mAP) = 0.902264, or 90.23 %

So model is good and training goes well. The issue is in the Training dataset.

----

> Also just as a note, since there are several negative images (where there are no lumps) from each video, I am doing a random sampling of negative images from hundreds of videos to gather equal number of images as positive images - as per your suggestion that I have read somewhere to have equal number of positive and negative images. So the dataset of 12000+ images here includes 6000+ positive images and 6000+ randomly sampled negative images. 

This is just a general recommendation. **But in your case with difficult to detect objects and very high False-Negatives (90%) - you should add much more Positive images.**

Just keep the same old data augmentation parameters jitter=0.3 and random=1, **collect Positive images as much as possible** (with your old Negative samples), and Train again with iteration number ~= number of images.


----

> Sure, I could do that...but I already saw overfitting happening around 4000 to 5000 iterations. For example the mAP that I have above is at just 2700 iterations and loss was showing up as 0.13+ which I believe is quite insignificant


Did you get the highest mAP for Training-set or for Test-set at 2700 iterations?
Usually the highest mAP for Training-set is achieved much earlier than for Test-set.

**Comment by kmsravindra** [comment]
@AlexeyAB , Thanks for your recommendations. I will try to collect more positive images.
